main/src/project.rs

```diff
@@ -228,6 +230,21 @@ pub fn extract_cargo_toml_version(cargo_toml_path: &PathBuf) -> Result<String> {
     Ok(version.to_string())
 }
 
+pub fn read_file_preimage(filename: &Path) -> Result<Vec<u8>> {
+    let mut contents = Vec::with_capacity(1024);
+    {
+        let filename = filename.as_os_str();
+        contents.extend_from_slice(&(filename.len() as u64).to_be_bytes());
+        contents.extend_from_slice(filename.as_encoded_bytes());
+    }
+    let mut file = std::fs::File::open(filename)
+        .map_err(|e| eyre!("failed to open file {}: {e}", filename.display()))?;
+    contents.extend_from_slice(&file.metadata().unwrap().len().to_be_bytes());
+    file.read_to_end(&mut contents)
```

Comment:
I'm really not 100% sure, but I think that memory mapping the file's contents, or using a buffered reader with large chunks should be more efficient that the default behavior, here.

Also, since these are almost always exclusively text files, I think that we don't have to worry too much about them being HUGE.

Reply:
Thanks for checking. Sorry about the wild goose chase.